subprojects/build-option/src/main/java/org/gradle/internal/buildoption/BuildOption.java

```diff
@@ -30,18 +30,42 @@
  */
 public interface BuildOption<T> {
 
-    @Nullable String getGradleProperty();
+    @Nullable
+    String getGradleProperty();
+
     void applyFromProperty(Map<String, String> properties, T settings);
+
     void configure(CommandLineParser parser);
+
     void applyFromCommandLine(ParsedCommandLine options, T settings);
 
     enum Origin {
-        GRADLE_PROPERTY, COMMAND_LINE;
+        GRADLE_PROPERTY {
+            @Override
+            public void handleInvalidValue(String propertyOrOption, String value, String hint) {
+                String message = String.format("Value '%s' given for %s Gradle property is invalid %s", value, propertyOrOption, hintMessage(hint));
+                throw new IllegalArgumentException(message);
+            }
+        },
+        COMMAND_LINE {
+            @Override
+            public void handleInvalidValue(String propertyOrOption, String value, String hint) {
+                String message = String.format("Argument value '%s' given for --%s option is invalid %s", value, propertyOrOption, hintMessage(hint));
+                throw new CommandLineArgumentException(message);
+            }
+        };
+
+        public abstract void handleInvalidValue(String propertyOrOption, String value, String hint);
```

Comment:
I don't think it's necessary to introduce another extra interface, could you explain it briefly?

Reply:
We might have another value handler implementation. In that case it would be easier to just use the interface in method signatures etc. It also makes testing easier.